File: code/java/164.java
class Solution {
    public int maximumGap(int[] nums) {
        int length = nums.length;
        if(length<=1) return 0;
        int minVal = Integer.MAX_VALUE;
        int maxVal = Integer.MIN_VALUE;
        for(int num:nums){
            if(num<minVal) minVal = num;
            if(num>maxVal) maxVal = num;
        }
        // 所有数字都一样
        if(maxVal==minVal) return 0;

        int interval = 0;
        // 间隔向上取整，保证length-1个桶可以包围[minVal,maxVal]这个范围
        if((float)(maxVal-minVal)/(length-1)>(maxVal-minVal)/(length-1)) interval = (maxVal-minVal)/(length-1)+1;
        else interval = (maxVal-minVal)/(length-1);

        int[] bucketMin = new int[length-1];
        int[] bucketMax = new int[length-1];
        for(int i=0;i<length-1;i++){
            bucketMax[i] = Integer.MIN_VALUE;
            bucketMin[i] = Integer.MAX_VALUE;
        }

        for(int num:nums){
            int index = (num-minVal)/interval;
            if(num==maxVal||num==minVal) continue;
            bucketMax[index] = Math.max(bucketMax[index],num);
            bucketMin[index] = Math.min(bucketMin[index],num);
        }

        int res = 0;
        int previous = minVal;
        for(int i=0;i<length-1;i++){
            if(bucketMax[i]==Integer.MIN_VALUE) continue;
            res = Math.max(bucketMin[i]-previous,res);
            previous = bucketMax[i];
        }
        res = Math.max(maxVal-previous,res);
        return res;
    }
}

File: code/python/164.py
class Solution:
    def maximumGap(self, nums: List[int]) -> int:
        if len(nums)<=1:
            return 0
        length = len(nums)
        minVal = min(nums)
        maxVal = max(nums)
        if maxVal-minVal==0:
            return 0
        #为了让桶的数量大于n-2,以n-1个桶来计算interval，保证一定有一个空的桶-->最大gap不会出现在桶内部的max-min中
        #每个箱子的范围
        if (maxVal-minVal)/(length-1)>(maxVal-minVal)//(length-1):
            interval = (maxVal-minVal)//(length-1)+1
        else:
            interval = (maxVal-minVal)//(length-1)
        #存放每个桶中的最大和最小值
        bucketMin = [float('inf')]*(length-1)
        bucketMax = [-1]*(length-1)

        for i in range(length):
            #计算nums[i]所在的桶的索引
            index = (nums[i]-minVal)//interval
            #不考虑最大和最小值
            if nums[i]==minVal or nums[i]==maxVal:
                continue
            #更新桶内的值
            bucketMin[index] = min(nums[i],bucketMin[index])
            bucketMax[index] = max(nums[i],bucketMax[index])

        print(interval)
        print(bucketMin)
        print(bucketMax)

        maxGap = 0
        #minVal视作第-1个箱子的最大值
        previousMax = minVal
        for i in range(length-1):
            if bucketMax[i]==-1:
                continue
            #当前桶的最小值减前一个桶的最大值
            maxGap = max(bucketMin[i]-previousMax,maxGap)
            previousMax = bucketMax[i]
        #maxVal-最后一桶的最大值
        maxGap = max(maxVal-previousMax,maxGap)
        return maxGap